server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,345 @@
+        logger.trace("createTempOutput() called {} , {}", prefix, suffix);
+        writeLock.lock();
+        try {
+            return localDirectory.createTempOutput(prefix, suffix, context);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Ensures that any writes to these files are moved to stable storage (made durable).
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void sync(Collection<String> names) throws IOException {
+        logger.trace("sync() called {}", names);
+        writeLock.lock();
+        try {
+            Collection<String> newLocalFiles = new ArrayList<>();
+            Collection<String> remoteFiles = Arrays.asList(getRemoteFiles());
+            for (String name : names) {
+                if (remoteFiles.contains(name) == false) newLocalFiles.add(name);
+            }
+            logger.trace("Synced files : {}", newLocalFiles);
+            localDirectory.sync(newLocalFiles);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Ensures that directory metadata, such as recent file renames, are moved to stable storage.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void syncMetaData() throws IOException {
+        logger.trace("syncMetaData() called ");
+        writeLock.lock();
+        try {
+            localDirectory.syncMetaData();
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Renames {@code source} file to {@code dest} file where {@code dest} must not already exist in
+     * the directory.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void rename(String source, String dest) throws IOException {
+        logger.trace("rename() called {}, {}", source, dest);
+        writeLock.lock();
+        try {
+            localDirectory.rename(source, dest);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Opens a stream for reading an existing file.
+     * Check whether the file is present locally or in remote and return the IndexInput accordingly
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public IndexInput openInput(String name, IOContext context) throws IOException {
+        logger.trace("openInput() called {}", name);
+        writeLock.lock();
+        try {
+            if (Arrays.asList(getRemoteFiles()).contains(name) == false) {
+                // If file has not yet been uploaded to Remote Store, fetch it from the local directory
+                logger.trace("File found in disk");
+                return localDirectory.openInput(name, context);
+            } else {
+                // If file has been uploaded to the Remote Store, fetch it from the Remote Store in blocks via
+                // OnDemandCompositeBlockIndexInput
```

Comment:
"This method must throw either NoSuchFileException or FileNotFoundException". How are we doing that by "fallback mechanism to OnDemandCompositeBlockIndexInput". 
This behavior is not inline with Lucene openInput specification in Directory abstraction (https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/store/Directory.java#L144)

Reply:
If file is not present in remote then we are falling back to the localDirectory which will take care of NoSuchFileException or FileNotFoundException.

When we are sure that the file is in remote, then only we are using OnDemandCompositeBlockIndexInput